Ở 2 chương trước, chúng ta đã biết vài cấu trúc Python cơ bản, đây là lúc chúng ta kết hợp những cấu trúc này để tạo ra những chương trình cơ bản. Và chúng ta sẽ đưa lên thêm vài cấu trúc và giải thuật.

# 3.1 Liệt kê kiệt sức (Exhaustive Enumeration)

Đoạn code ở dưới dùng để tính căn bậc 3 (nếu tồn tại) của một số nguyên. Nếu không tồn tại nó sẽ in ra rằng số đưa vào không phải là bậc 3 của một số nguyên _(not a perfect cube)_. 

In [8]:
#Find the cube root of a perfect cube
x = int(input('Enter a integer:'))
# For example: 7406961012236344616
ans = 0
while ans**3 <abs(x):
    ans = ans + 1
if ans**3 != abs(x):
    print(x, ' is not a perfect cube')
else:
    if x < 0:
        ans = -ans
    print('Cube root of',x , 'is', ans)

Enter a integer:7406961012236344616
Cube root of 7406961012236344616 is 1949306


Với giá trị nào của x thì chương trình dừng lại? Câu trả lời là "với bất kỳ số nào". Lý do rất đơn giản:
- Giá trị của ans\*\*3 bắt đầu bằng 0, và trở nên lớn hơn sau các vòng lặp.
- Khi nó vượt quá abs(x), vòng lặp dừng lại.
- Vì abs(x) là giá trị tuyệt đối của x, nó luôn dương nên chỉ có một số lần lặp hữu hạn trước khi vòng lặp phải kết thúc.

Khi viết vòng lặp, phải nghĩ về **hàm giảm (decremeting function)**. Hàm này phải có tính chất:
- Nó ánh xạ một tập hợp các biến chương trình thành một số nguyên.
- Khi vào vòng lặp, giá trị của nó là không âm.
- Khi giá trị của nó là $\leq$ 0, vòng lặp chấm dứt.
- Giá trị của nó bị giảm mỗi lần qua vòng lặp.

Hàm giảm của chương trình trên là gì? đó là: **abs(x) - ans\*\*3**

Giải thuật ở trên là một biến thể của **đoán và kiểm tra (guess and check)** được gọi là **exhaustive enumeration**. Chúng ta liệt kê tất cả các khả năng cho đến khi chúng ta có được câu trả lời. Khi mới nhìn, chúng ta thấy nó thật là ngu ngốc nhưng đây là giải thuật thực tiễn nhất để giải quyết một vấn đề. Và trong nhiều trường hợp, nó vẫn đủ nhanh cho các mục đích thực tiễn, dù cho 1 triệu lần đoán dược yêu cầu, cũng không gây ra vấn đề gì.

In [11]:
#just for fun
maxVal = int(input('Enter a postive integer: '))
i = 0
while i < maxVal:
    i = i + 1
    print(i)

Enter a postive integer: 8
1
2
3
4
5
6
7
8


# 3.2 Vòng lặp for (for loop)

Vòng lặp for có cấu trúc như ở dưới:
```python
for variable in sequence:
    code block
```

Chuỗi các giá trị được cột với biến được tạo phổ biến nhất bằng cách sử dụng **hàm range**, trả về một chuỗi các số nguyên.  Hàm range sẽ bắt đầu bằng start, **start + step, start + 2\*step... start + i\*step, stop**. Step có thể âm hoặc dương. Xem ví dụ ở dưới: start = 0, stop = x, step = 1

In [13]:
x = 4
for i in range(0, x):
    print(i)

0
1
2
3


Chúng ta thử nghĩ về đoạn code này: 

In [14]:
x = 4
for i in range(0, x):
    print(i)
    x = 5

0
1
2
3


Khi gán lại giá trị cho x trong vòng lặp for liệu nó có ảnh hưởng tới iterator? Câu trả lời là không. Chúng ta xem thêm 2 đoạn code dưới: 

In [16]:
x = 4
for j in range(x):
    for i in range(x):
        print(i)
        #x = 2 #không gán lại x

0
1
2
3
0
1
2
3
0
1
2
3
0
1
2
3


In [18]:
x = 4
for j in range(x):
    for i in range(x):
        print(i)
        x = 2 #gán lại x

0
1
2
3
0
1
0
1
0
1


Vì hàm range ở vòng lặp ngoài chỉ được gán một lần, còn hàm range ở vòng lặp trong sẽ được gán mỗi lẫn vòng for ở trong chạy.

Đoạn code ở dưới được viết lại dùng vòng lặp for và break để tìm căn bậc 3: 

In [20]:
#Find the cube root of a perfect cube
x = int(input('Enter an integer: '))
for ans in range(0, abs(x)+1):
    if ans**3 >= abs(x):
        break
if ans**3 != abs(x):
    print(x, 'is not a perfect cube')
else:
    if x < 0:
        ans = -ans
    print('Cube root of', x,'is', ans)

Enter an integer: 65
65 is not a perfect cube


Vòng lặp for có thể được dùng để kết hợp với toán tử **in**, chúng ta xem ví dụ ở dưới:

In [22]:
total = 0
for c in '12345678':
    print(c)
    total = total + int(c)
print(total)

1
2
3
4
5
6
7
8
36


# 3.3 Phương pháp xấp xỉ và phương pháp chia đôi

Giả sử ai đó yêu cầu bạn viết chương trình tính căn bậc 2 của một số không âm, giả sử cân bậc 2 của 2 chẳng hạn. Chúng ta biết căn 2 không phải là số hữu tỉ, nên không thể tính được giá trị chính xác của số này. 
<br/> Yêu cầu đúng phải là viết chương trình tìm **xấp xỉ (approximation)** của căn bậc 2 của 2. 

Đoạn code ở dưới dùng một giải thuật đơn giản để tìm một xấp xỉ của căn bậc 2. 
<br/> Lưu ý: *toán tử += là cách viết rút gọn, ans = ans + step có thể được viết rút gọn là ans += step.*

In [3]:
x = 25
epsilon = 0.01
step = epsilon**2
numGuess = 0
ans = 0.0

while abs(ans**2 - x) >= epsilon and ans <= x:
    ans += step
    numGuess += 1    
print ('numGuess =', numGuess)

if abs(ans**2 - x) >= epsilon:
    print('Failed on square root of', x)
else:
    print(ans, 'is close to square root of', x)

numGuess = 49990
4.999000000001688 is close to square root of 25


Một lần nữa, chúng ta đã sử dụng _exhaustive enumeration_.
<br/>Với x = 25, chúng ta có kết quả:
<br/>numGuesses = 49990
<br/>4.999000000001688 is close to square root of 25

In [9]:
x = 0.25 #thử lại giá trị x mới là 0.25
epsilon = 0.01
step = epsilon**2
numGuess = 0
ans = 0.0

while abs(ans**2 - x) >= epsilon and ans <= x:
    ans += step
    numGuess += 1    
print ('numGuess =', numGuess)

if abs(ans**2 - x) >= epsilon:
    print('Failed on square root of', x)
else:
    print(ans, 'is close to square root of', x)

numGuess = 2501
Failed on square root of 0.25


Khi thử lại với x = 0.25, chúng ta có:
<br/>numGuesses = 2501
<br/>Failed on square root of 0.25

Với code ở trên chúng ta liệt kê từ 0 đến x, nhưng khi x từ 0 tới 1, thì **căn bậc 2 của x lại lớn hơn x** nên không nằm trên đoạn \[0,1\]này. 
<br/> Một cách để sửa lại sai sót này là sửa điều kiện vòng lặp while lại thành:  
<br/>while abs(ans\*\*2 - x) >= epsilon and **ans\*ans <= x**:

In [11]:
x = 0.25
epsilon = 0.01
step = epsilon**2
numGuess = 0
ans = 0.0

while abs(ans**2 - x) >= epsilon and ans*ans <= x: #update điều kiện chỗ này
    ans += step
    numGuess += 1    
print ('numGuess =', numGuess)

if abs(ans**2 - x) >= epsilon:
    print('Failed on square root of', x)
else:
    print(ans, 'is close to square root of', x)

numGuess = 4899
0.48989999999996237 is close to square root of 0.25


Chúng ta lại thử lại với x = 123456

In [10]:
x = 123456 #thử lại giá trị x mới 123456
epsilon = 0.01
step = epsilon**2
numGuess = 0
ans = 0.0

while abs(ans**2 - x) >= epsilon and ans*ans <= x: 
    ans += step
    numGuess += 1    
print ('numGuess =', numGuess)

if abs(ans**2 - x) >= epsilon:
    print('Failed on square root of', x)
else:
    print(ans, 'is close to square root of', x)

numGuess = 3513631
Failed on square root of 123456


Chương trình chạy hơi lâu và chúng ta có kết quả là:
<br/> numGuess = 3513631
<br/> Failed on square root of 123456
<br/>Chuyện gì đã xảy ra, rõ ràng tồn tại một số xấp xỉ căn của 123456 với sai số 0.01. Tại sao chương trình không tính ra?
<br/> Đó là do chúng ta chọn step quá lớn, chương trình sẽ bỏ quả các kết quả hợp lí. Chúng ta thử lại với **step = epsilon\*\*3**:

In [14]:
x = 123456
epsilon = 0.01
step = epsilon**3 #update step chỗ này
numGuess = 0
ans = 0.0

while abs(ans**2 - x) >= epsilon and ans*ans <= x:
    ans += step
    numGuess += 1    
print ('numGuess =', numGuess)

if abs(ans**2 - x) >= epsilon:
    print('Failed on square root of', x)
else:
    print(ans, 'is close to square root of', x)

numGuess = 351363047
351.36304620491023 is close to square root of 123456


Tuy cuối cùng cũng ra kết quả, nhưng chúng ta phải chờ rất lâu, chương trình cần **hơn 351 TRIỆU** lần đoán để ra kết quả.
<br/> Rõ ràng tới đây chúng ta cần cách khác để giải quyết vấn đề, chúng ta cần giải thuật tốt hơn là tinh chỉnh step.
<br/> Chúng ta cần tiếp cận vấn đề theo một cách khác. Giả sử chúng ta biết rằng, xấp xỉ căn bậc 2 nằm đâu đó ở đoạn 0 - max, chúng ta bắt đầu đoán ở giữa đoạn này, nếu đây không phải là câu trả lời đúng thì liệu rằng nó quá lớn hay quá nhỏ. Nếu quá nhỏ, chúng ta biết kết quả nằm ở nửa bên phải và ngược lại. Chúng ta chọn nửa đoạn đó thành đoạn mới và chọn tiếp một giá trị ở giữa. Rồi chúng ta tiếp tục cho tới khi tới kết quả.
<br/> Ở dưới là đoạn code mô tả giải thuật này:

In [15]:
x = 25
epsilon = 0.01
numGuesses = 0
low = 0.0
high = max(1.0, x)
ans = (high + low)/2.0

while abs(ans**2 - x) >= epsilon:
    print('low =', low, 'high =', high, 'ans =', ans)
    numGuesses += 1
    if ans**2 < x:
        low = ans
    else:
        high = ans
    ans = (high + low)/2.0
print('numGuesses =', numGuesses)
print(ans, 'is close to square root of', x)

low = 0.0 high = 25 ans = 12.5
low = 0.0 high = 12.5 ans = 6.25
low = 0.0 high = 6.25 ans = 3.125
low = 3.125 high = 6.25 ans = 4.6875
low = 4.6875 high = 6.25 ans = 5.46875
low = 4.6875 high = 5.46875 ans = 5.078125
low = 4.6875 high = 5.078125 ans = 4.8828125
low = 4.8828125 high = 5.078125 ans = 4.98046875
low = 4.98046875 high = 5.078125 ans = 5.029296875
low = 4.98046875 high = 5.029296875 ans = 5.0048828125
low = 4.98046875 high = 5.0048828125 ans = 4.99267578125
low = 4.99267578125 high = 5.0048828125 ans = 4.998779296875
low = 4.998779296875 high = 5.0048828125 ans = 5.0018310546875
numGuesses = 13
5.00030517578125 is close to square root of 25


In [2]:
x = 123456 #thử lại với 123456
epsilon = 0.01
numGuesses = 0
low = 0.0
high = max(1.0, x)
ans = (high + low)/2.0

while abs(ans**2 - x) >= epsilon:
    print('low =', low, 'high =', high, 'ans =', ans)
    numGuesses += 1
    if ans**2 < x:
        low = ans
    else:
        high = ans
    ans = (high + low)/2.0
print('numGuesses =', numGuesses)
print(ans, 'is close to square root of', x)

low = 0.0 high = 123456 ans = 61728.0
low = 0.0 high = 61728.0 ans = 30864.0
low = 0.0 high = 30864.0 ans = 15432.0
low = 0.0 high = 15432.0 ans = 7716.0
low = 0.0 high = 7716.0 ans = 3858.0
low = 0.0 high = 3858.0 ans = 1929.0
low = 0.0 high = 1929.0 ans = 964.5
low = 0.0 high = 964.5 ans = 482.25
low = 0.0 high = 482.25 ans = 241.125
low = 241.125 high = 482.25 ans = 361.6875
low = 241.125 high = 361.6875 ans = 301.40625
low = 301.40625 high = 361.6875 ans = 331.546875
low = 331.546875 high = 361.6875 ans = 346.6171875
low = 346.6171875 high = 361.6875 ans = 354.15234375
low = 346.6171875 high = 354.15234375 ans = 350.384765625
low = 350.384765625 high = 354.15234375 ans = 352.2685546875
low = 350.384765625 high = 352.2685546875 ans = 351.32666015625
low = 351.32666015625 high = 352.2685546875 ans = 351.797607421875
low = 351.32666015625 high = 351.797607421875 ans = 351.5621337890625
low = 351.32666015625 high = 351.5621337890625 ans = 351.44439697265625
low = 351.32666015625 high =

Phương pháp này gọi là **tìm kiếm chia đôi (bisection search)**, giải thuật này tốt hơn hẳn giải thuật trước.
<br/> Tuy nhiên, nếu thay 25 bằng -25, nó sẽ tạo ra vòng lặp vô hạn.

# 3.4 Vài lời bàn về kiểu float

Hầu hết mọi lúc, các số kiểu float cung cấp xấp xỉ hợp lý so với số thực. Nhưng "hầu hết" không có nghĩa là "luôn luôn". Chúng ta xem ví dụ dưới:

In [3]:
x = 0.0
for i in range(10):
    x = x + 0.1
if x == 1.0:
    print(x, '= 1.0')
else:
    print(x, 'is not 1.0')

0.9999999999999999 is not 1.0


Để hiểu vấn đề này, chúng ta phải hiểu *số có dấu chấm động (floating point number)* được lưu trữ như thế nào trong máy tính trong suốt quá trình tính toán, và để hiểu điều này chúng ta phải hiểu *số nhị phân (binary number)*.
<br/> Số nhị phân là những số có cơ số 2, chỉ biểu diễn 0 hoặc 1 gọi là *bit*. Ví dụ: 101 ở cơ số 2 sẽ là 5 ở cơ số 10.

Trong hầu hết ngôn ngữ lập trình, số không nguyên (non-integer) sẽ được biểu diễn dưới dạng **dấu chấm động (floating point)**. Chúng gồm 1 cặp số nguyên: **phần trọng số (significant digit)** và **phần mũ (exponenet)**. 
<br/>Ví dụ: 1.949 sẽ được lưu thành 1 cặp (1949, -3), tức là $1949*10^{-3}$.
<br/> Phần trọng số sẽ quyết định sự chính xác (precision) của số được lưu. Trường hợp trọng số chỉ được lưu bởi 2 số thì 1.949 sẽ trở thành 1.9, sự xấp xỉ này được gọi là **làm tròn (rounded value)**

Máy tính hiện đại sử dụng số nhị phân, không phải thập phân như ở trên. 
<br/> Ví dụ số 0.625 (5/8) sẽ được biểu diễn bởi cặp (101,-11) vì  101 ở nhị phân  là 5 ở thập phân và -11 ở nhị phân là -3 ở thập phân, do đó cặp (101, -11) sẽ biểu diễn cho $5*2^{-3} = 5/8 = 0.625$

Nếu biểu diễn bằng 4 số nhị phân, thì kết quả tốt nhất chúng ta có là (0011, -101), cặp này có giá trị $3*2^{-5} = 3/32 = 0.09375$
<br/> Nếu biểu diễn bằng 5 số nhị phân, thì kết quả tốt nhất chúng ta có là (11001, -1000), cặp này có giá trị $25*2^{-8} = 25/256 = 0.09765625$
<br/> Nói tóm lại, chúng sẽ được biểu diễn bằng một cặp (sig, exp) có giá trị $sig*2^{exp}$ nhưng tiếc thay không tồn tại cặp nào có giá trị đúng bằng 0.1
<br/> Trong Python, sẽ có 53 bit để lưu phần trọng số, và với số thập phân 0.1 thì phần trọng số được lưu là:
<br/>11001100110011001100110011001100110011001100110011001
và có giá trị tương đương số thập phân là:
<br/>0.1000000000000000055511151231257827021181583404541015625
rất gần 0.1 nhưng không phải là 0.1

Tiện thể thì có hàm **round(x, numDigits)** để làm tròn số có dấu chấm động. 

In [4]:
print(round(2**0.5, 3))
print(round(2**0.5, 5))
print(round(2**0.5, 10))

1.414
1.41421
1.4142135624


Vậy việc này có gây ảnh hưởng gì không? 
<br/> Đa phần thì không, nhưng nhiều khi có nhiều lỗi bất ngờ xảy ra.
<br/> Viết code **abs(x-y) /< 0.0001** thì tốt hơn là **x == y**

# 3.5 Giải thuật Newton-Raphson

Newton (và Raphson) đã chứng mình một định lý rằng, một giá trị guess xấp xỉ với *nghiệm (root)* của *đa thức p(x) (polynomial)* thì giá trị 
<br/>guess – $\frac{p(guess)} {p’(guess)}$ sẽ có một giá trị xấp xỉ tốt hơn với p' là đạo hàm của p.

Ở phía trên chúng ta muốn tính căn bậc 2 của một số, thì rõ ràng đa thức này có dạng $x^2-k = 0$, với k không âm và đạo hàm của đa thức này là $2x$. 
<br/>Giả sử *giá trị đoán hiện tại (current guess)* là y, thì *giá trị đoán kế tiếp (next guess)* là $y-(y^2-k)/2y$, điều này gọi là **xấp xỉ liên tiếp (successive approximation)**.

In [13]:
#Newton-Raphson for square root
#Find x such that x**2 - 24 is within epsilon of 0.01
epsilon = 0.01
k = 123456
step = 0
guess = k/2.0
while abs(guess*guess - k) >= epsilon:
    guess = guess - (((guess**2) - k)/(2*guess))
    step += 1
    
print('numGuesses =',step)    
print('Square root of', k, 'is about', guess)

numGuesses = 11
Square root of 123456 is about 351.3630601485824
